packages/playwright-core/src/server/android/android.ts

```diff
@@ -256,15 +260,23 @@ export class AndroidDevice extends SdkObject {
     await androidBrowser._init();
     this._browserConnections.add(androidBrowser);
 
+    const artifactsDir = await fs.promises.mkdtemp(ARTIFACTS_FOLDER);
+    socket.on('close', async () => {
+      const errors = await removeFolders([artifactsDir]);
+      if (errors) {
+        for (let i = 0; i < errors.length; ++i)
+          debug('pw:android')(`exception while removing ${artifactsDir}: ${errors[i]}`);
+      }
+    });
     const browserOptions: BrowserOptions = {
       ...this._android._playwrightOptions,
       name: 'clank',
       isChromium: true,
       slowMo: 0,
       persistent: { ...options, noDefaultViewport: true },
-      artifactsDir: '',
-      downloadsPath: '',
-      tracesDir: '',
+      artifactsDir,
+      downloadsPath: artifactsDir,
+      tracesDir: artifactsDir,
```

Comment:
- I think this fixes https://github.com/microsoft/playwright/issues/12214!
- While you are here, perhaps fix the TODO at line 294?